### PR TITLE
Rewrite media-embed to typescript.

### LINKED ISSUE #13023: [TS] Rewrite ckeditor5-media-embed into TypeScript
Part of #11704.

### PR DESCRIPTION
**Pull request by mremiszewski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (media-embed): Rewrite `ckeditor5-media-embed` to Typescript. Closes #13023.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-media-embed/src/mediaembed.ts [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+ */
+
+import { Plugin, type PluginDependencies } from 'ckeditor5/src/core';
+import { Widget } from 'ckeditor5/src/widget';
+
+import MediaEmbedEditing from './mediaembedediting';
+import AutoMediaEmbed from './automediaembed';
+import MediaEmbedUI from './mediaembedui';
+
+import '../theme/mediaembed.css';
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembed.ts:17]: CSS should be imported at the end of imports block.

File: packages/ckeditor5-media-embed/src/automediaembed.ts [hunk 2]
```diff
@@ -22,61 +22,56 @@ const URL_REGEXP = /^(?:http(s)?:\/\/)?[\w-]+\.[\w-.~:/?#[\]@!$&'()*+,;=%]+$/;
...
-	static get requires() {
+	public static get requires(): PluginDependencies {
 		return [ Clipboard, Delete, Undo ];
 	}
 
 	/**
 	 * @inheritDoc
 	 */
-	static get pluginName() {
+	public static get pluginName(): 'AutoMediaEmbed' {
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/automediaembed.ts:37]: Please augment `PluginsMap`.

File: packages/ckeditor5-media-embed/src/mediaembedcommand.ts [hunk 3]
```diff
@@ -60,28 +67,25 @@ export default class MediaEmbedCommand extends Command {
...
 }
 
-// Checks if the table is allowed in the parent.
-//
-// @param {module:engine/model/selection~Selection|module:engine/model/documentselection~DocumentSelection} selection
-// @param {module:engine/model/model~Model} model
-// @returns {Boolean}
-function isAllowedInParent( selection, model ) {
+/**
+ * Checks if the table is allowed in the parent.
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembedcommand.ts:71]:
```suggestion
 * Checks if the media embed is allowed in the parent.
```

File: packages/ckeditor5-media-embed/src/mediaembedediting.ts [hunk 4]
```diff
@@ -7,35 +7,40 @@
...
+import { first, type GetCallback } from 'ckeditor5/src/utils';
 
 import { modelToViewUrlAttributeConverter } from './converters';
 import MediaEmbedCommand from './mediaembedcommand';
 import MediaRegistry from './mediaregistry';
 import { toMediaWidget, createMediaFigureElement } from './utils';
 
 import '../theme/mediaembedediting.css';
+import type { MediaEmbedConfig } from './mediaembed';
+import type { UpcastElementEvent } from 'ckeditor5/src/engine';
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembedediting.ts:20]: Please keep imports from `ckeditor5/src/...` next to one another.

File: packages/ckeditor5-media-embed/src/mediaembedediting.ts [hunk 5]
```diff
@@ -7,35 +7,40 @@
...
+	public registry: MediaRegistry;
+
 	/**
 	 * @inheritDoc
 	 */
-	constructor( editor ) {
+	constructor( editor: Editor ) {
 		super( editor );
-
+		// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembedediting.ts:43]: Why is this needed?

File: packages/ckeditor5-media-embed/src/mediaembedediting.ts [hunk 6]
```diff
@@ -268,13 +270,25 @@ export default class MediaEmbedEditing extends Plugin {
...
 
-					const modelElement = first( modelRange.getItems() );
+					const modelElement = first( modelRange!.getItems() );
 
 					if ( !modelElement ) {
 						// Revert consumed figure so other features can convert it.
 						conversionApi.consumable.revert( data.viewItem, { name: true, classes: 'media' } );
 					}
-				}
+				};
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembedediting.ts:279]:
```suggestion
				};

```

File: packages/ckeditor5-media-embed/src/mediaembedtoolbar.ts [hunk 7]
```diff
@@ -51,11 +48,22 @@ export default class MediaEmbedToolbar extends Plugin {
...
- * @member {Array.<String>} module:media-embed/mediaembed~MediaEmbedConfig#toolbar
- */
+declare module '@ckeditor/ckeditor5-core' {
+	interface PluginsMap {
+		[ MediaEmbedToolbar.pluginName ]: MediaEmbedToolbar;
+	}
+}
+
+declare module './mediaembed'
+{
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembedtoolbar.ts:58]:
```suggestion
declare module './mediaembed' {
```

File: packages/ckeditor5-media-embed/src/mediaembedui.ts [hunk 8]
```diff
@@ -95,10 +91,10 @@ export default class MediaEmbedUI extends Plugin {
 			} );
 
 			form.delegate( 'submit', 'cancel' ).to( dropdown );
-			form.urlInputView.bind( 'value' ).to( command, 'value' );
+			form.urlInputView.fieldView.bind( 'value' ).to( command, 'value' );
 
 			// Form elements should be read-only when corresponding commands are disabled.
-			form.urlInputView.bind( 'isReadOnly' ).to( command, 'isEnabled', value => !value );
+			form.urlInputView.bind( 'isEnabled' ).to( command, 'isEnabled' );
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaembedui.ts:97]: I would expect that this code worked previously so please make sure that this change is needed.

File: packages/ckeditor5-media-embed/src/mediaregistry.ts [hunk 9]
```diff
@@ -85,26 +86,25 @@ export default class MediaRegistry {
...
 	 * Returns a `Media` instance for the given URL.
 	 *
 	 * @protected
-	 * @param {String} url The URL of the media.
-	 * @returns {module:media-embed/mediaregistry~Media|null} The `Media` instance or `null` when there is none.
+	 * @param url The URL of the media.
+	 * @returns The `Media` instance or `null` when there is none.
 	 */
-	_getMedia( url ) {
+	protected _getMedia( url: string ): Media | null {
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaregistry.ts:107]: Is this method used anywhere in overriding class? If not maybe it should be `private`?

File: packages/ckeditor5-media-embed/src/mediaregistry.ts [hunk 10]
```diff
@@ -171,49 +170,46 @@ export default class MediaRegistry {
  * @private
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaregistry.ts:170]: This is not needed. This class is not exported.

File: packages/ckeditor5-media-embed/src/mediaregistry.ts [hunk 11]
```diff
@@ -171,49 +170,46 @@ export default class MediaRegistry {
...
-		this._match = match;
+	/**
+	 * The output of the `RegExp.match` which validated the {@link #url} of this media.
+	 */
+	private _match?: object;
+
+	/**
+	 * The function returning the HTML string preview of this media.
+	 */
+	private _previewRenderer?: Function;
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/mediaregistry.ts:194]: Can we provide some more detailed type (not `Function`)?

File: packages/ckeditor5-media-embed/src/utils.ts [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+ */
+export function createMediaFigureElement(
+	writer: DowncastWriter,
+	registry: MediaRegistry,
+	url: string,
+	options: MediaOptions
+): ViewContainerElement {
+	return writer.createContainerElement( 'figure', { class: 'media' }, [
+		registry.getMediaViewElement( writer, url, options ),
+		writer.createSlot( 'children' )
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/utils.ts:83]:
```suggestion
		writer.createSlot()
```

File: packages/ckeditor5-media-embed/src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+			findOptimalPosition: findOptimalPosition ? 'auto' : undefined
+		} );
+	} );
+}
+
+export type MediaOptions = {
+	elementName: string;
+	renderMediaPreview?: boolean;
+	renderForEditingView?: boolean;
+};
```
Comment (niegowski) [inline packages/ckeditor5-media-embed/src/utils.ts:125]: Missing docs.

**Review by niegowski — CHANGES_REQUESTED** [review]
Also missing changes to `.gitignore` file.

File: packages/ckeditor5-media-embed/src/mediaembedui.ts (hunk 8 above)
Comment (niegowski) [reply packages/ckeditor5-media-embed/src/mediaembedui.ts:97]: Ok I checked it and it looks like it was invalid before (probably while we were introducing labeled fields).

File: packages/ckeditor5-media-embed/src/mediaembedediting.ts (hunk 5 above)
Comment (mremiszewski) [reply packages/ckeditor5-media-embed/src/mediaembedediting.ts:43]: I forgot to remove this line.

File: packages/ckeditor5-media-embed/src/mediaembed.ts [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+	 */
+	previewsInData?: boolean;
+}
+
+declare module '@ckeditor/ckeditor5-core' {
+	interface PluginsMap {
+		[ MediaEmbed.pluginName ]: MediaEmbed;
+	}
+
+	interface EditorConfig {
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/mediaembed.ts:324]: Move this declaration and config-related interfaces to `mediaembedconfig.ts`

File: packages/ckeditor5-media-embed/src/mediaembed.ts [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+	 * {@link module:media-embed/mediaembed~MediaEmbedConfig#providers `config.mediaEmbed.providers`} and
+	 * {@link module:media-embed/mediaembed~MediaEmbedConfig#extraProviders `config.mediaEmbed.extraProviders`}
+	 *
+	 * ```ts
+	 * mediaEmbed: {
+	 * 	removeProviders: [ 'youtube', 'twitter' ]
+	 * }
+	 * ```
+	 */
+	removeProviders: Array<string>;
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/mediaembed.ts:250]: I think it should be nullable.

File: packages/ckeditor5-media-embed/src/mediaembedediting.ts [hunk 16]
```diff
@@ -268,13 +269,26 @@ export default class MediaEmbedEditing extends Plugin {
...
 			} );
 	}
 }
+
+declare module '@ckeditor/ckeditor5-core' {
+	interface PluginsMap {
+		[ MediaEmbedEditing.pluginName ]: MediaEmbedEditing;
+	}
+
+	interface CommandsMap {
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/mediaembedediting.ts:290]: Move it to `mediaembedcommand.ts`.

File: packages/ckeditor5-media-embed/src/mediaembedtoolbar.ts [hunk 17]
```diff
@@ -51,11 +48,21 @@ export default class MediaEmbedToolbar extends Plugin {
...
- * @member {Array.<String>} module:media-embed/mediaembed~MediaEmbedConfig#toolbar
- */
+declare module '@ckeditor/ckeditor5-core' {
+	interface PluginsMap {
+		[ MediaEmbedToolbar.pluginName ]: MediaEmbedToolbar;
+	}
+}
+
+declare module './mediaembed' {
+	interface MediaEmbedConfig {
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/mediaembedtoolbar.ts:58]: Merge this with `MediaEmbedConfig` declared in `mediaembedconfig.ts`.

File: packages/ckeditor5-media-embed/src/mediaregistry.ts [hunk 18]
```diff
@@ -167,53 +164,48 @@ export default class MediaRegistry {
...
-		 * @see module:utils/locale~Locale#t
-		 * @method
-		 */
-		this._locale = locale;
+	/**
+	 * Shorthand for {@link module:utils/locale~Locale#t}.
+	 *
+	 * @see module:utils/locale~Locale#t
+	 * @method
+	 */
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/mediaregistry.ts:179]:
```suggestion
	 */
```

File: packages/ckeditor5-media-embed/src/ui/mediaformview.ts [hunk 19]
```diff
@@ -29,82 +29,90 @@ import '../../theme/mediaform.css';
...
+	 * Tracks information about the DOM focus in the form.
+	 *
+	 * @readonly
+	 */
+	public readonly focusTracker: FocusTracker;
+
+	/**
+	 * An instance of the {@link module:utils/keystrokehandler~KeystrokeHandler}.
+	 *
+	 * @readonly
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/ui/mediaformview.ts:44]: Unnecessary.

File: packages/ckeditor5-media-embed/src/ui/mediaformview.ts [hunk 20]
```diff
@@ -29,82 +29,90 @@ import '../../theme/mediaform.css';
...
- *
- * @extends module:ui/view~View
  */
 export default class MediaFormView extends View {
 	/**
-	 * @param {Array.<Function>} validators Form validators used by {@link #isValid}.
-	 * @param {module:utils/locale~Locale} [locale] The localization services instance.
+	 * Tracks information about the DOM focus in the form.
+	 *
+	 * @readonly
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/ui/mediaformview.ts:37]: Unnecessary.

File: packages/ckeditor5-media-embed/src/ui/mediaformview.ts [hunk 21]
```diff
@@ -342,11 +319,11 @@ export default class MediaFormView extends View {
  * Fired when the form view is submitted (when one of the children triggered the submit event),
  * e.g. click on {@link #saveButtonView}.
  *
- * @event submit
+ * @eventName submit
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/ui/mediaformview.ts:322]: `eventName` should be attached to a type describing this event. It seems not used in this code anyway.

File: packages/ckeditor5-media-embed/src/ui/mediaformview.ts [hunk 22]
```diff
@@ -342,11 +319,11 @@ export default class MediaFormView extends View {
...
  */
 
 /**
  * Fired when the form view is canceled, e.g. by a click on {@link #cancelButtonView}.
  *
- * @event cancel
+ * @eventName cancel
```
Comment (arkflpc) [inline packages/ckeditor5-media-embed/src/ui/mediaformview.ts:328]: `eventName` should be attached to a type describing this event. It seems not used in this code anyway.

**Review by arkflpc — CHANGES_REQUESTED** [review]

**Review by arkflpc — APPROVED** [review]

**Pull request merged**